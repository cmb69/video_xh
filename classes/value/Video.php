<?php

/**
 * Copyright 2012-2023 Christoph M. Becker
 *
 * This file is part of Video_XH.
 *
 * Video_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Video_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Video_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Video\Value;

class Video
{
    /** @var array<string,string> */
    private $sources;

    /** @var string|null */
    private $poster;

    /** @var string|null */
    private $subtitle;

    /** @var int */
    private $date;

    /** @param array<string,string> $sources */
    public function __construct(array $sources, ?string $poster, ?string $subtitle, int $date)
    {
        $this->sources = $sources;
        $this->poster = $poster;
        $this->subtitle = $subtitle;
        $this->date = $date;
    }

    public function filename(): string
    {
        assert(!empty($this->sources));
        return key($this->sources);
    }

    /** @return array<string,string> */
    public function sources(): array
    {
        return $this->sources;
    }

    public function poster(): ?string
    {
        return $this->poster;
    }

    public function subtitle(): ?string
    {
        return $this->subtitle;
    }

    public function date(): int
    {
        return $this->date;
    }
}
