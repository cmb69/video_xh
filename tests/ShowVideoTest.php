<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Video_XH.
 *
 * Video_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Video_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Video_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Video;

use function XH_includeVar;
use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use Plib\FakeRequest;
use Plib\View;
use Video\Value\Video;
use Video\Infra\VideoFinder;
use Video\Logic\OptionParser;

class ShowVideoTest extends TestCase
{
    private const OPTIONS = "title=My%20Video&description=This%20is%20a%20nice%20one%20%26%20it%27s%20short";

    /** @var ShowVideo */
    private $sut;

    /** @var VideoFinder&MockObject */
    private $videoFinder;

    public function setUp(): void
    {
        $this->videoFinder = $this->createStub(VideoFinder::class);
        $this->sut = new ShowVideo(
            new OptionParser(XH_includeVar("./config/config.php", "plugin_cf")['video']),
            $this->videoFinder,
            new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")['video'])
        );
    }

    public function testRendersVideoWithPoster(): void
    {
        $this->videoFinder->method('find')->willReturn(new Video(
            [
                './userfiles/media/my_video.mp4' => "mp4",
                './userfiles/media/my_video.webm' => "webm",
            ],
            "./userfiles/media/my_video.jpg",
            null,
            1674668829
        ));
        $response = ($this->sut)(new FakeRequest(["language" => "en"]), "my_video", self::OPTIONS);
        Approvals::verifyHtml($response->output());
    }

    public function testRendersVideoWithoutPoster(): void
    {
        $this->videoFinder->method('find')->willReturn(new Video(
            [
                './userfiles/media/my_video.mp4' => "mp4",
                './userfiles/media/my_video.webm' => "webm",
            ],
            null,
            null,
            1674668829
        ));
        $response = ($this->sut)(new FakeRequest(["language" => "en"]), "my_video", self::OPTIONS);
        Approvals::verifyHtml($response->output());
    }

    public function testReportsMissingVideo(): void
    {
        $response = ($this->sut)(new FakeRequest(["language" => "en"]), "no_video", self::OPTIONS);
        Approvals::verifyHtml($response->output());
    }
}
