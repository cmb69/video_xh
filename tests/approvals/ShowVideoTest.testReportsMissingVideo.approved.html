<p class="xh_fail">Video &quot;no_video&quot; missing!</p>